File: tools/binding_compliance/conformance/families/mod_ini.py
"""Facts for actual typed cache getters and complete read-only mod INI scans."""

from collections.abc import Mapping
from functools import partial
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy
from .file_operations import _files


def _observed(operation: str, observation: Mapping[str, Any]) -> bool:
    """Require exact file preservation and every projected native result field."""
    if (
        set(observation) != {"operation", "before", "result", "files"}
        or observation["operation"] != operation
        or not _files(observation["files"])
        or observation["before"] != observation["files"]
    ):
        return False
    value = observation["result"]
    if not isinstance(value, Mapping):
        return False
    if operation == "cache":
        return (
            set(value)
            == {
                "names",
                "contains",
                "path",
                "name",
                "enabled",
                "count",
                "scale",
                "hasName",
                "missing",
                "duplicates",
            }
            and isinstance(value["names"], list)
            and value["names"] == sorted(set(value["names"]))
            and value["contains"] is ("sample.ini" in value["names"])
            and (value["path"] is not None) == value["contains"]
            and value["hasName"] is (value["name"] is not None)
            and value["missing"] is None
            and isinstance(value["duplicates"], Mapping)
            and (value["enabled"] is None or type(value["enabled"]) is bool)
            and (value["count"] is None or type(value["count"]) is int)
            and (value["scale"] is None or isinstance(value["scale"], str))
        )
    return (
        set(value) == {"message", "issues", "vsync", "duplicates"}
        and isinstance(value["message"], str)
        and all(
            isinstance(value[key], list) for key in ("issues", "vsync", "duplicates")
        )
        and all(
            isinstance(entry, Mapping)
            and set(entry) == {"path", "setting"}
            and all(isinstance(item, str) for item in entry.values())
            for entry in value["vsync"]
        )
    )


def _vsync(observation: Mapping[str, Any]) -> bool:
    """Only a nonempty native VSync list proves construction of its entry carrier."""
    return _observed("scan", observation) and bool(observation["result"]["vsync"])


def _duplicates(observation: Mapping[str, Any]) -> bool:
    """Require published groups/maps to agree before and after a real file replacement."""
    if (
        set(observation) != {"operation", "before", "result", "files"}
        or observation["operation"] != "duplicates"
        or not _files(observation["before"])
        or not _files(observation["files"])
    ):
        return False
    before = {item["path"]: item["content"] for item in observation["before"]}
    after = {item["path"]: item["content"] for item in observation["files"]}
    value = observation["result"]
    if (
        set(before) != set(after)
        or sum(before[path] != after[path] for path in before) != 1
        or not isinstance(value, Mapping)
        or set(value) != {"initialGroups", "initialMap", "afterGroups", "afterMap"}
    ):
        return False
    groups = value["initialGroups"]
    if (
        not isinstance(groups, list)
        or not groups
        or value["afterGroups"] != []
        or value["afterMap"] != {}
    ):
        return False
    expected = {}
    for group in groups:
        if (
            not isinstance(group, Mapping)
            or set(group) != {"original", "duplicates"}
            or not isinstance(group["original"], str)
            or not isinstance(group["duplicates"], list)
            or not group["duplicates"]
        ):
            return False
        paths = [group["original"], *group["duplicates"]]
        if (
            not all(path in before for path in paths)
            or len(paths) != len(set(paths))
            or any(before[path] != before[paths[0]] for path in paths)
        ):
            return False
        expected[group["original"].split("/")[-1].lower()] = paths
    return value["initialMap"] == expected


MOD_INI_COVERAGE_POLICY = FamilyCoveragePolicy(
    "mod-ini",
    (
        CoveragePredicate(
            "duplicate-lifecycle",
            "mod-ini.duplicates",
            "mod-ini.duplicates",
            "duplicate-groups",
            ("ConfigDuplicateDetector", "DuplicateGroup"),
            _duplicates,
            runtime_operations=(
                None,
                "__init__",
                "detect_duplicates",
                "get_duplicate_map",
                "detect_config_duplicates",
                "detectConfigDuplicates",
            ),
        ),
        CoveragePredicate(
            "typed-cache",
            "mod-ini.cache",
            "mod-ini.cache",
            "ini-cache",
            ("ConfigFileCache",),
            partial(_observed, "cache"),
            runtime_operations=(
                None,
                *(
                    f"RustConfigFileCache.{name}"
                    for name in (
                        "__init__",
                        "config_files",
                        "contains",
                        "get_bool",
                        "get_duplicates",
                        "get_float",
                        "get_int",
                        "get_path",
                        "get_str",
                        "has_setting",
                    )
                ),
            ),
        ),
        CoveragePredicate(
            "mod-scan",
            "mod-ini.scan",
            "mod-ini.scan",
            "mod-scan",
            ("ModIniScanner", "scan", "ModIniScanResult"),
            partial(_observed, "scan"),
            runtime_operations=(
                None,
                "RustModIniScanner.__init__",
                "RustModIniScanner.scan",
                "scan_mod_inis",
                "scanModInis",
            ),
        ),
        CoveragePredicate(
            "vsync-entry",
            "mod-ini.scan",
            "mod-ini.scan",
            "mod-scan",
            ("VsyncEntry",),
            _vsync,
            runtime_operations=(None,),
        ),
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/mod_ini.rs
//! Native typed INI cache and mod scan observations over owned fixture files.

use super::file_operations::{destination, files};
use super::{RunnerResult, invalid, text};
use classic_scangame_core::{config_cache::ConfigFileCache, mod_ini::ModIniScanner};
use serde_json::{Value, json};
use std::{collections::BTreeMap, fs, path::Path};

/// Converts an actual native path to a contained portable fixture path.
fn relative(root: &Path, path: &Path) -> RunnerResult<String> {
    Ok(path
        .strip_prefix(root)?
        .to_string_lossy()
        .replace('\\', "/"))
}

/// Executes public native cache getters or the full mod scanner without mutating inputs.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    let temporary = tempfile::tempdir()?;
    let root = temporary.path();
    for (path, content) in fixture["files"]
        .as_object()
        .ok_or_else(|| invalid("files must be an object"))?
    {
        let target = destination(root, path)?;
        fs::create_dir_all(
            target
                .parent()
                .ok_or_else(|| invalid("file needs parent"))?,
        )?;
        fs::write(target, text(content)?.as_bytes())?;
    }
    let operation = text(&fixture["operation"])?;
    let mut result = json!({"operation":operation,"before":files(root)?});
    result["result"] = if operation == "cache" {
        let mut cache = ConfigFileCache::new(root, &[])?;
        let mut names = cache.config_files().keys().cloned().collect::<Vec<_>>();
        names.sort();
        let path = cache
            .get_path("sample.ini")
            .map(|path| relative(root, path))
            .transpose()?;
        let mut duplicates = BTreeMap::new();
        for (name, paths) in &cache.duplicate_files {
            let mut paths = paths
                .iter()
                .map(|path| relative(root, path))
                .collect::<RunnerResult<Vec<_>>>()?;
            paths.sort();
            duplicates.insert(name.clone(), paths);
        }
        json!({"names":names,"contains":cache.contains("sample.ini"),"path":path,
            "name":cache.get_str("sample.ini","Settings","name"),"enabled":cache.get_bool("sample.ini","Settings","enabled"),
            "count":cache.get_int("sample.ini","Settings","count"),"scale":cache.get_float("sample.ini","Settings","scale").map(|value|format!("{value:.3}")),
            "hasName":cache.has_setting("sample.ini","Settings","name"),"missing":cache.get_str("absent.ini","Settings","name"),"duplicates":duplicates})
    } else if operation == "duplicates" {
        use classic_scangame_core::ConfigDuplicateDetector;
        let project = |detector: &ConfigDuplicateDetector| -> RunnerResult<Vec<Value>> {
            let mut groups = detector
                .get_duplicates()
                .values()
                .map(|group| {
                    let mut copies = group
                        .duplicates
                        .iter()
                        .map(|path| relative(root, path))
                        .collect::<RunnerResult<Vec<_>>>()?;
                    copies.sort();
                    Ok(json!({"original":relative(root,&group.canonical)?,"duplicates":copies}))
                })
                .collect::<RunnerResult<Vec<_>>>()?;
            groups
                .sort_by(|left, right| left["original"].as_str().cmp(&right["original"].as_str()));
            Ok(groups)
        };
        let mapping = |values: std::collections::HashMap<String, Vec<std::path::PathBuf>>| -> RunnerResult<BTreeMap<String, Vec<String>>>{ values.into_iter().map(|(name, paths)| Ok((name, paths.iter().map(|path| relative(root, path)).collect::<RunnerResult<Vec<_>>>()?))).collect() };
        let mut detector = ConfigDuplicateDetector::new();
        let initial_map = mapping(detector.scan_directory(root)?)?;
        let initial = project(&detector)?;
        let mut excluded =
            ConfigDuplicateDetector::with_whitelist(vec!["no-fixture-matches".into()]);
        if !excluded.scan_directory(root)?.is_empty() {
            return Err(invalid("custom duplicate whitelist ignored").into());
        }
        let replacement = destination(root, &text(&fixture["replacement"]["path"])?)?;
        fs::write(
            replacement,
            text(&fixture["replacement"]["content"])?.as_bytes(),
        )?;
        let after_map = mapping(detector.scan_directory(root)?)?;
        let after = project(&detector)?;
        json!({"initialGroups":initial,"initialMap":initial_map,"afterGroups":after,"afterMap":after_map})
    } else if operation == "scan" {
        let value = ModIniScanner::scan(root, &text(&fixture["game"])?)?;
        let issues = value.issues.iter().map(|issue| Ok(json!({"filePath":relative(root,&issue.file_path)?,"section":issue.section,"setting":issue.setting,"currentValue":issue.current_value,"recommendedValue":issue.recommended_value,"description":issue.description,"severity":format!("{:?}",issue.severity)}))).collect::<RunnerResult<Vec<_>>>()?;
        let vsync = value
            .vsync_files
            .iter()
            .map(|entry| {
                Ok(json!({"path":relative(root,&entry.file_path)?,"setting":entry.setting}))
            })
            .collect::<RunnerResult<Vec<_>>>()?;
        let duplicates = value
            .duplicates
            .iter()
            .map(|entry| {
                let mut paths = entry
                    .paths
                    .iter()
                    .map(|path| relative(root, path))
                    .collect::<RunnerResult<Vec<_>>>()?;
                paths.sort();
                Ok(json!({"name":entry.file_name,"paths":paths}))
            })
            .collect::<RunnerResult<Vec<_>>>()?;
        json!({"message":value.message.replace(&*root.to_string_lossy(),"<ROOT>").replace('\\',"/"),"issues":issues,"vsync":vsync,"duplicates":duplicates})
    } else {
        return Err(invalid("unsupported mod INI operation").into());
    };
    result["files"] = files(root)?;
    Ok(result)
}
